src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3_inpaint.py

```diff
@@ -0,0 +1,1033 @@
+            height,
+            width,
+            prompt_embeds.dtype,
+            device,
+            generator,
+        )
+
+        # 6. Denoising loop
+        num_warmup_steps = max(len(timesteps) - num_inference_steps * self.scheduler.order, 0)
+        self._num_timesteps = len(timesteps)
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                if self.interrupt:
+                    continue
+
+                # expand the latents if we are doing classifier free guidance
+                latent_model_input = torch.cat([latents] * 2) if self.do_classifier_free_guidance else latents
+                # broadcast to batch dimension in a way that's compatible with ONNX/Core ML
+                timestep = t.expand(latent_model_input.shape[0])
+
+                noise_pred = self.transformer(
+                    hidden_states=latent_model_input,
+                    timestep=timestep,
+                    encoder_hidden_states=prompt_embeds,
+                    pooled_projections=pooled_prompt_embeds,
+                    return_dict=False,
+                )[0]
+
+                # perform guidance
+                if self.do_classifier_free_guidance:
+                    noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
+                    noise_pred = noise_pred_uncond + self.guidance_scale * (noise_pred_text - noise_pred_uncond)
+
+                # compute the previous noisy sample x_t -> x_t-1
+                latents_dtype = latents.dtype
+                latents = self.scheduler.step(noise_pred, t, latents, return_dict=False)[0]
+
+                if latents.dtype != latents_dtype:
+                    if torch.backends.mps.is_available():
+                        # some platforms (eg. apple mps) misbehave due to a pytorch bug: https://github.com/pytorch/pytorch/pull/99272
+                        latents = latents.to(latents_dtype)
+
+                if callback_on_step_end is not None:
+                    callback_kwargs = {}
+                    for k in callback_on_step_end_tensor_inputs:
+                        callback_kwargs[k] = locals()[k]
+                    callback_outputs = callback_on_step_end(self, i, t, callback_kwargs)
+
+                    latents = callback_outputs.pop("latents", latents)
+                    prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+                    negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
+                    negative_pooled_prompt_embeds = callback_outputs.pop(
+                        "negative_pooled_prompt_embeds", negative_pooled_prompt_embeds
+                    )
+
+                if add_predicted_noise:
+                    init_latents_proper = self.scheduler.scale_noise(
+                        init_latents_orig, torch.tensor([t]), noise_pred_uncond
+                    )
+                else:
```

Comment:
can we match the SD and SDXL inpainting pipeline when using the regular unet checkpoint https://github.com/huggingface/diffusers/blob/d5dd8df3b4e978e3f0549fa35e245a8785116af7/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py#L1272

Reply:
Thanks. I updated it with checking the number of transformer channels before running the denoising process.

Reply:
sorry I wasn't clear

When I said "can we match the SD and SDXL inpainting pipeline when using the regular unet checkpoint", I didn't mean that we need to check the number of transformer channels, although it is ok if you add the check.

SD/SDXL inpainting pipelines support both inpainting-specific checkpoints (when num_channels_unet==9) and regular text-to-image checkpoint (when num_channels_unet =4); I think the algorithm and overall code structure should match very closely with SD/SDXL inpainting, but you can ignore the part of logic in these pipeline that only that applies to  inpainting-specific checkpoints.

the current implementation of this pipeline matches the inpainting_legacy, which we deprecated and is slightly different from SDXL and SD, both in code structure and the actual algorithm